source/rust_verify/src/verus_items.rs

```diff
@@ -0,0 +1,638 @@
+        TyKind::Uint(t) => format!("{}", t.name_str()),
+        TyKind::Float(t) => format!("{}", t.name_str()),
+        TyKind::RawPtr(ref tm) => format!(
+            "*{} {}",
+            match tm.mutbl {
+                rustc_ast::Mutability::Mut => "mut",
+                rustc_ast::Mutability::Not => "const",
+            },
+            ty_to_stable_string_partial(tcx, &tm.ty)?.into_inner(),
+        ),
+        TyKind::Ref(_r, ty, mutbl) => format!(
+            "&{} {}",
+            match mutbl {
+                rustc_ast::Mutability::Mut => "mut",
+                rustc_ast::Mutability::Not => "const",
+            },
+            ty_to_stable_string_partial(tcx, ty)?.into_inner(),
+        ),
+        TyKind::Never => format!("!"),
+        TyKind::Tuple(ref tys) => format!(
+            "({})",
+            tys.iter()
+                .map(|ty| ty_to_stable_string_partial(tcx, &ty).map(|x| x.into_inner()))
+                .collect::<Option<Vec<_>>>()?
+                .join(",")
+        ),
+        TyKind::Param(ref param_ty) => format!("{}", param_ty.name.as_str()),
+        TyKind::Adt(def, _substs) => {
+            return Some(TyToStableStringPartialResult::Path(def_id_to_stable_rust_path(
+                tcx,
+                def.did(),
+            )?));
+        }
+        TyKind::Str => format!("str"),
+        TyKind::Array(ty, sz) => {
+            format!("[{}; {}]", ty_to_stable_string_partial(tcx, &ty)?.into_inner(), sz)
+        }
+        TyKind::Slice(ty) => format!("[{}]", ty_to_stable_string_partial(tcx, &ty)?.into_inner()),
+        _ => return None,
+    }))
+}
+
+pub(crate) fn def_id_to_stable_rust_path<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> Option<String> {
+    let def_path = tcx.def_path(def_id);
+    let mut segments: Vec<String> = Vec::with_capacity(def_path.data.len());
+    let crate_name = tcx.crate_name(def_path.krate);
+    segments.push(crate_name.to_ident_string());
+    let mut one_impl_block_in_path = false;
+    for d in def_path.data.iter() {
+        use rustc_hir::definitions::DefPathData;
+        match &d.data {
+            DefPathData::ValueNs(symbol) | DefPathData::TypeNs(symbol) => {
+                segments.push(symbol.to_string())
+            }
+            DefPathData::Ctor => segments.push(vir::def::RUST_DEF_CTOR.to_string()),
+            DefPathData::Impl => {
+                if one_impl_block_in_path {
+                    return None;
+                }
+                one_impl_block_in_path = true;
+                let self_ty = tcx.type_of(tcx.parent(def_id));
+                match ty_to_stable_string_partial(tcx, &self_ty)? {
+                    TyToStableStringPartialResult::Path(path) => {
+                        segments.clear();
+                        segments.push(path);
+                    }
+                    TyToStableStringPartialResult::Segment(segment) => {
+                        segments.push(segment);
+                    }
```

Comment:
I don't totally get the logic here - if the type is something like `bool`, then we get a 'Segment', and if the type is an ADT we get a `Path` ... why do those types have different behavior? I'd have expected we'd want to do TYPE_PATH::method_name in either case.

Reply:
Because the ADT printer will print its full path, so we replace the path we've seen so far with the ADT's full path. This is similar to what we do for friendly names, but more focused to produce stable identifier.

Reply:
I'm not sure I understand the question, but let me be more precise, what I said before wasn't quite right: when we print the path for a impl function, we want to print the "self" type path, and not the path with the disambiguator (`impl%0`), so when I reach an `impl` in the path I call out to the printer function recursively, but having replaced the `def_id` with that of the the "self" type. When that happens, we want to throw away the path constructed so far, and use the one for the "self" type instead.

Reply:
the TyToStableStringPartialResult::Path case - which does what you just described - makes sense to me. I'm confused about the TyToStableStringPartialResult::Segment case, which doesn't clear the path.

Reply:
Ah! Good catch. That case doesn't really make sense if we want to stay close to the rust paths.
I had a bit of a broken model of what was happening because the paths I saw made sense (`core::num::u64::MAX` vs `u64::MAX`), so I didn't see what you meant initially.

Reply:
Fixed in the commit I just pushed. Feel free to un-resolve if something still doesn't make sense.

Reply:
Ok, the path construction makes more sense now, but now that it's changed, I'm wondering, is `u8` properly unambiguous on its own? Maybe ty_to_stable_string_partial should return `"core::num::u64"`?

Reply:
Why would it be ambiguous? i.e. what could it collide with? (I think `core::num::u64` is the deprecated path, see https://doc.rust-lang.org/std/u64/constant.MAX.html)

Reply:
oh, i see. i'm not really sure, then

Reply:
wait, actually, i do think it's possible for users to have crates called 'bool' or 'u8' or something.

is there a check that the rust item really lives in core or std?